mergify_engine/context.py

```diff
@@ -56,6 +57,17 @@
 MARKDOWN_COMMIT_MESSAGE_RE = re.compile(r"^#+ Commit Message ?:?\s*$", re.I)
 
 
+# NOTE(sileht): Sentinel object (eg: `marker = object()`) can't be expressed
+# with typing yet use the proposed workaround instead:
+#   https://github.com/python/typing/issues/689
+#   https://www.python.org/dev/peps/pep-0661/
+class _NotCached(enum.Enum):
+    _MARKER = 0
+
+
+NotCached: typing.Final = _NotCached._MARKER
```

Comment:
We're using this a lot, I'm wondering if having a general `Unset` marker in a util module wouldn't be simpler.

Reply:
Each thing should have it's marker in theory to avoid bug like:

```
def bar(o = Unset):
      if o is Unset:
          print("I'm unset)

def foo(o = Unset):
      bar(o)

>>> foo()
I'm unset
```
o has been set for bar, but detected as unset.

Reply:
I'm not sure it's a bug in this case. The alternative would be:
```python
def bar(o=UnsetBar):
   if o is UnsetBar:
       print("I'm unset")

def foo(o=UnsetFoo):
    bar(o)
```

Except that calling `bar` with `UnsetFoo` as value has no sense at all, so it'd likely be a bug.

Your example has semantic sense, mine does not.
Your example works as expected, mine would probably raise an error.

Reply:
> Except that calling bar with UnsetFoo as value has no sense at all, so it'd likely be a bug.

A bug cached by mypy

Reply:
> A bug cached by mypy

Right, but it's still a bug. Why do you want to have a bug by default? 🤔

Reply:
I'm just saying in my example, you also have the bug, but mypy will not complain as type match.

If `foo()` and `bar()` play well together and are related why not sharing the `Unset` as the meaning/compat is also shared. 

I don't like copy/pasting the enum hack for marker and hope mypy implement it nicely.

Reply:
That's my point: `Unset` always have the same meaning, e.g., `Unset` argument in this case.
That's why it makes total sense to use only one `Unset` type.

That being said I think your example is not well chosen for this case here.